radicle-tracker/src/thread.rs

```diff
@@ -0,0 +1,162 @@
+use rose_tree::{NodeIndex, RoseTree, ROOT};
+use std::{collections::HashMap, hash};
+
+/// Laws:
+/// Thread::new(comment).first() === comment
+/// thread.reply(comment).delete(comment) === thread
+/// Thread::new(comment).delete(comment) === None
+/// Thread::new(comment).edit(f, comment) === Thread::new(f(comment).unwrap())
+#[derive(Debug, Clone)]
+pub struct Thread<A> {
+    lut: HashMap<Path, NodeIndex>,
+    tree: RoseTree<A>,
+}
+
+#[derive(Debug, PartialEq, Eq, hash::Hash, Clone)]
+pub struct Path(Vec<u32>);
+
+impl Path {
+    pub fn new(index: u32) -> Self {
+        Path(vec![index])
+    }
+
+    pub fn push(&mut self, index: u32) {
+        self.0.push(index);
+    }
+}
+
+impl<A> Thread<A> {
+    pub fn new(a: A) -> (Self, Path)
+    where
+        A: Eq,
+    {
+        let (tree, root) = RoseTree::new(a);
+        let mut lut = HashMap::new();
+        let path = Path::new(0);
+        lut.insert(path.clone(), root);
+        (Thread { lut, tree }, path)
+    }
+
+    pub fn delete(&mut self, path: &Path) -> Result<A, ()> {
+        match self.lut.remove(&path) {
+            Some(ix) => self.tree.remove_node(ix).ok_or(()),
+            None => Err(()),
+        }
+    }
+
+    pub fn edit<F>(&mut self, path: &Path, f: F) -> Option<A>
+    where
+        F: FnOnce(&A) -> Option<A>,
```

Comment:
Shouldn't this take a `&mut A`?

Reply:
I thought that at first, but because in the future there may be some ACL semantics, the function should not be able to mutate the `A`.

Unless... the functions hides this away anyway and mutates the `A` if the user accessing is allowed otherwise don't mutate it at all. God mutation semantics are so dirty :grimacing:

Reply:
Oh, but then why isn't `self` mutable? Guess I'm confused :grimacing:

Reply:
Because if we get Some(new_edit) then you can replace the old value.

Would you say is more Rustier?

Reply:
I'm probably missing something -- maybe a usage example would help? I just don't understand how you can use this function to mutate the value :confused: